// @aries-framework/anoncreds/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.storeLinkSecret = exports.assertBestPracticeRevocationInterval = exports.generateLegacyProverDidLikeString = exports.AnonCredsApi = exports.AnonCredsModuleConfig = exports.AnonCredsModule = void 0;
require("reflect-metadata");
__exportStar(require("./models"), exports);
__exportStar(require("./services"), exports);
__exportStar(require("./error"), exports);
__exportStar(require("./repository"), exports);
__exportStar(require("./formats"), exports);
__exportStar(require("./protocols"), exports);
var AnonCredsModule_1 = require("./AnonCredsModule");
Object.defineProperty(exports, "AnonCredsModule", { enumerable: true, get: function () { return AnonCredsModule_1.AnonCredsModule; } });
var AnonCredsModuleConfig_1 = require("./AnonCredsModuleConfig");
Object.defineProperty(exports, "AnonCredsModuleConfig", { enumerable: true, get: function () { return AnonCredsModuleConfig_1.AnonCredsModuleConfig; } });
var AnonCredsApi_1 = require("./AnonCredsApi");
Object.defineProperty(exports, "AnonCredsApi", { enumerable: true, get: function () { return AnonCredsApi_1.AnonCredsApi; } });
__exportStar(require("./AnonCredsApiOptions"), exports);
var proverDid_1 = require("./utils/proverDid");
Object.defineProperty(exports, "generateLegacyProverDidLikeString", { enumerable: true, get: function () { return proverDid_1.generateLegacyProverDidLikeString; } });
__exportStar(require("./utils/indyIdentifiers"), exports);
var revocationInterval_1 = require("./utils/revocationInterval");
Object.defineProperty(exports, "assertBestPracticeRevocationInterval", { enumerable: true, get: function () { return revocationInterval_1.assertBestPracticeRevocationInterval; } });
var linkSecret_1 = require("./utils/linkSecret");
Object.defineProperty(exports, "storeLinkSecret", { enumerable: true, get: function () { return linkSecret_1.storeLinkSecret; } });
