// @hyperledger/indy-vdr-nodejs/build/ffi/primitives.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFI_POOL_HANDLE = exports.FFI_REQUEST_HANDLE_POINTER = exports.FFI_REQUEST_HANDLE = exports.FFI_HANDLE_POINTER = exports.FFI_HANDLE = exports.FFI_ERROR_CODE = exports.FFI_CALLBACK_PTR = exports.FFI_CALLBACK_ID = exports.FFI_INT32_PTR = exports.FFI_STRING_PTR = exports.FFI_INT8_PTR = exports.FFI_POINTER = exports.FFI_VOID = exports.FFI_STRING = exports.FFI_INT64 = exports.FFI_INT32 = exports.FFI_INT8 = exports.FFI_USIZE = exports.FFI_UINT64 = exports.FFI_UINT8 = void 0;
const ref_napi_1 = require("@2060.io/ref-napi");
// Primitives
exports.FFI_UINT8 = 'uint8';
exports.FFI_UINT64 = 'uint64';
exports.FFI_USIZE = 'size_t';
exports.FFI_INT8 = 'int8';
exports.FFI_INT32 = 'int32';
exports.FFI_INT64 = 'int64';
exports.FFI_STRING = 'string';
exports.FFI_VOID = 'void';
exports.FFI_POINTER = 'pointer';
// Pointers
exports.FFI_INT8_PTR = (0, ref_napi_1.refType)(exports.FFI_INT8);
exports.FFI_STRING_PTR = (0, ref_napi_1.refType)(exports.FFI_STRING);
exports.FFI_INT32_PTR = (0, ref_napi_1.refType)(exports.FFI_INT32);
// Custom
exports.FFI_CALLBACK_ID = exports.FFI_INT64;
exports.FFI_CALLBACK_PTR = exports.FFI_POINTER;
exports.FFI_ERROR_CODE = exports.FFI_INT64;
exports.FFI_HANDLE = 'int64';
exports.FFI_HANDLE_POINTER = 'int64*';
exports.FFI_REQUEST_HANDLE = exports.FFI_HANDLE;
exports.FFI_REQUEST_HANDLE_POINTER = exports.FFI_HANDLE_POINTER;
exports.FFI_POOL_HANDLE = exports.FFI_HANDLE;
