// @hyperledger/anoncreds-nodejs/build/library/bindings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nativeBindings = void 0;
const ffi_1 = require("../ffi");
exports.nativeBindings = {
    // first element is method return type, second element is list of method argument types
    anoncreds_buffer_free: [ffi_1.FFI_VOID, [ffi_1.ByteBufferStruct]],
    anoncreds_create_credential: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_create_credential_definition: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_INT8,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_create_credential_offer: [
        ffi_1.FFI_ERRORCODE,
        [ffi_1.FFI_STRING, ffi_1.FFI_STRING, ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_OBJECT_HANDLE_PTR],
    ],
    anoncreds_create_credential_request: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_create_link_secret: [ffi_1.FFI_ERRORCODE, [ffi_1.FFI_OBJECT_HANDLE_PTR]],
    anoncreds_create_or_update_revocation_state: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_INT64,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_create_presentation: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.CredentialEntryListStruct,
            ffi_1.CredentialProveListStruct,
            ffi_1.StringListStruct,
            ffi_1.StringListStruct,
            ffi_1.FFI_STRING,
            ffi_1.ObjectHandleListStruct,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleListStruct,
            ffi_1.StringListStruct,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_create_revocation_registry_def: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_STRING,
            ffi_1.FFI_INT64,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_create_schema: [
        ffi_1.FFI_ERRORCODE,
        [ffi_1.FFI_STRING, ffi_1.FFI_STRING, ffi_1.FFI_STRING, ffi_1.StringListStruct, ffi_1.FFI_OBJECT_HANDLE_PTR],
    ],
    anoncreds_credential_get_attribute: [ffi_1.FFI_ERRORCODE, [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, ffi_1.FFI_STRING_PTR]],
    anoncreds_encode_credential_attributes: [ffi_1.FFI_ERRORCODE, [ffi_1.StringListStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_generate_nonce: [ffi_1.FFI_ERRORCODE, [ffi_1.FFI_STRING_PTR]],
    anoncreds_get_current_error: [ffi_1.FFI_ERRORCODE, [ffi_1.FFI_STRING_PTR]],
    anoncreds_object_free: [ffi_1.FFI_VOID, [ffi_1.FFI_OBJECT_HANDLE]],
    anoncreds_string_free: [ffi_1.FFI_VOID, [ffi_1.FFI_STRING_PTR]],
    anoncreds_object_get_json: [ffi_1.FFI_ERRORCODE, [ffi_1.FFI_OBJECT_HANDLE, ffi_1.ByteBufferStructPtr]],
    anoncreds_object_get_type_name: [ffi_1.FFI_ERRORCODE, [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING_PTR]],
    anoncreds_presentation_request_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_process_credential: [
        ffi_1.FFI_ERRORCODE,
        [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_OBJECT_HANDLE_PTR],
    ],
    anoncreds_revocation_registry_definition_get_attribute: [
        ffi_1.FFI_ERRORCODE,
        [ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_STRING, ffi_1.FFI_STRING_PTR],
    ],
    anoncreds_set_default_logger: [ffi_1.FFI_ERRORCODE, []],
    anoncreds_verify_presentation: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.ObjectHandleListStruct,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleListStruct,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleListStruct,
            ffi_1.StringListStruct,
            ffi_1.ObjectHandleListStruct,
            ffi_1.NonRevokedIntervalOverrideListStruct,
            ffi_1.FFI_INT8_PTR,
        ],
    ],
    anoncreds_create_revocation_status_list: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_STRING,
            ffi_1.FFI_INT8,
            ffi_1.FFI_INT64,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_update_revocation_status_list_timestamp_only: [
        ffi_1.FFI_ERRORCODE,
        [ffi_1.FFI_INT64, ffi_1.FFI_OBJECT_HANDLE, ffi_1.FFI_OBJECT_HANDLE_PTR],
    ],
    anoncreds_update_revocation_status_list: [
        ffi_1.FFI_ERRORCODE,
        [
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.FFI_OBJECT_HANDLE,
            ffi_1.I32ListStruct,
            ffi_1.I32ListStruct,
            ffi_1.FFI_INT64,
            ffi_1.FFI_OBJECT_HANDLE_PTR,
        ],
    ],
    anoncreds_version: [ffi_1.FFI_STRING, []],
    anoncreds_credential_request_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_credential_request_metadata_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_presentation_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_credential_offer_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_revocation_registry_definition_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_revocation_registry_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_revocation_status_list_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_revocation_state_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_credential_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_credential_definition_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_credential_definition_private_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_revocation_registry_definition_private_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_key_correctness_proof_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
    anoncreds_schema_from_json: [ffi_1.FFI_ERRORCODE, [ffi_1.ByteBufferStruct, ffi_1.FFI_STRING_PTR]],
};
