// @hyperledger/anoncreds-nodejs/build/ffi/primitives.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FFI_STRING_PTR = exports.FFI_OBJECT_HANDLE_PTR = exports.FFI_INT8_PTR = exports.FFI_ISIZE_PTR = exports.FFI_STRING = exports.FFI_VOID = exports.FFI_OBJECT_HANDLE = exports.FFI_ERRORCODE = exports.FFI_UINT8 = exports.FFI_UINT = exports.FFI_INT64 = exports.FFI_INT32 = exports.FFI_INT8 = exports.FFI_ISIZE = void 0;
const ref_napi_1 = require("@2060.io/ref-napi");
// Primitives
exports.FFI_ISIZE = 'size_t';
exports.FFI_INT8 = 'int8';
exports.FFI_INT32 = 'int32';
exports.FFI_INT64 = 'int64';
exports.FFI_UINT = 'uint';
exports.FFI_UINT8 = 'uint8';
exports.FFI_ERRORCODE = exports.FFI_UINT;
exports.FFI_OBJECT_HANDLE = exports.FFI_ISIZE;
exports.FFI_VOID = ref_napi_1.types.void;
exports.FFI_STRING = 'string';
// Pointers
exports.FFI_ISIZE_PTR = (0, ref_napi_1.refType)(exports.FFI_ISIZE);
exports.FFI_INT8_PTR = (0, ref_napi_1.refType)(exports.FFI_INT8);
exports.FFI_OBJECT_HANDLE_PTR = (0, ref_napi_1.refType)(exports.FFI_OBJECT_HANDLE);
exports.FFI_STRING_PTR = (0, ref_napi_1.refType)(exports.FFI_STRING);
