// @aries-framework/anoncreds/build/utils/indyIdentifiers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseIndyRevocationRegistryId = exports.parseIndyCredentialDefinitionId = exports.parseIndySchemaId = exports.parseIndyDid = exports.isDidIndyRevocationRegistryId = exports.isDidIndyCredentialDefinitionId = exports.isDidIndySchemaId = exports.isUnqualifiedSchemaId = exports.isUnqualifiedRevocationRegistryId = exports.isUnqualifiedCredentialDefinitionId = exports.getUnqualifiedRevocationRegistryId = exports.getUnqualifiedCredentialDefinitionId = exports.getUnqualifiedSchemaId = exports.didIndyRegex = exports.didIndyRevocationRegistryIdRegex = exports.unqualifiedRevocationRegistryIdRegex = exports.didIndyCredentialDefinitionIdRegex = exports.unqualifiedCredentialDefinitionIdRegex = exports.unqualifiedIndyDidRegex = exports.unqualifiedSchemaVersionRegex = exports.didIndySchemaIdRegex = exports.unqualifiedSchemaIdRegex = void 0;
const core_1 = require("@aries-framework/core");
const didIndyAnonCredsBase = /(did:indy:((?:[a-z][_a-z0-9-]*)(?::[a-z][_a-z0-9-]*)?):([1-9A-HJ-NP-Za-km-z]{21,22}))\/anoncreds\/v0/;
// <namespaceIdentifier>:2:<schemaName>:<schemaVersion>
exports.unqualifiedSchemaIdRegex = /^([a-zA-Z0-9]{21,22}):2:(.+):([0-9.]+)$/;
// did:indy:<namespace>:<namespaceIdentifier>/anoncreds/v0/SCHEMA/<schemaName>/<schemaVersion>
exports.didIndySchemaIdRegex = new RegExp(`^${didIndyAnonCredsBase.source}/SCHEMA/(.+)/([0-9.]+)$`);
exports.unqualifiedSchemaVersionRegex = /^(\d+\.)?(\d+\.)?(\*|\d+)$/;
exports.unqualifiedIndyDidRegex = /^(did:sov:)?[a-zA-Z0-9]{21,22}$/;
// <namespaceIdentifier>:3:CL:<schemaSeqNo>:<tag>
exports.unqualifiedCredentialDefinitionIdRegex = /^([a-zA-Z0-9]{21,22}):3:CL:([1-9][0-9]*):(.+)$/;
// did:indy:<namespace>:<namespaceIdentifier>/anoncreds/v0/CLAIM_DEF/<schemaSeqNo>/<tag>
exports.didIndyCredentialDefinitionIdRegex = new RegExp(`^${didIndyAnonCredsBase.source}/CLAIM_DEF/([1-9][0-9]*)/(.+)$`);
// <namespaceIdentifier>:4:<schemaSeqNo>:3:CL:<credentialDefinitionTag>:CL_ACCUM:<revocationRegistryTag>
exports.unqualifiedRevocationRegistryIdRegex = /^([a-zA-Z0-9]{21,22}):4:[a-zA-Z0-9]{21,22}:3:CL:([1-9][0-9]*):(.+):CL_ACCUM:(.+)$/;
// did:indy:<namespace>:<namespaceIdentifier>/anoncreds/v0/REV_REG_DEF/<schemaSeqNo>/<credentialDefinitionTag>/<revocationRegistryTag>
exports.didIndyRevocationRegistryIdRegex = new RegExp(`^${didIndyAnonCredsBase.source}/REV_REG_DEF/([1-9][0-9]*)/(.+)/(.+)$`);
exports.didIndyRegex = /^did:indy:((?:[a-z][_a-z0-9-]*)(?::[a-z][_a-z0-9-]*)?):([1-9A-HJ-NP-Za-km-z]{21,22})$/;
function getUnqualifiedSchemaId(unqualifiedDid, name, version) {
    return `${unqualifiedDid}:2:${name}:${version}`;
}
exports.getUnqualifiedSchemaId = getUnqualifiedSchemaId;
function getUnqualifiedCredentialDefinitionId(unqualifiedDid, seqNo, tag) {
    return `${unqualifiedDid}:3:CL:${seqNo}:${tag}`;
}
exports.getUnqualifiedCredentialDefinitionId = getUnqualifiedCredentialDefinitionId;
// TZQuLp43UcYTdtc3HewcDz:4:TZQuLp43UcYTdtc3HewcDz:3:CL:98158:BaustellenzertifikateNU1:CL_ACCUM:1-100
function getUnqualifiedRevocationRegistryId(unqualifiedDid, seqNo, credentialDefinitionTag, revocationRegistryTag) {
    return `${unqualifiedDid}:4:${unqualifiedDid}:3:CL:${seqNo}:${credentialDefinitionTag}:CL_ACCUM:${revocationRegistryTag}`;
}
exports.getUnqualifiedRevocationRegistryId = getUnqualifiedRevocationRegistryId;
function isUnqualifiedCredentialDefinitionId(credentialDefinitionId) {
    return exports.unqualifiedCredentialDefinitionIdRegex.test(credentialDefinitionId);
}
exports.isUnqualifiedCredentialDefinitionId = isUnqualifiedCredentialDefinitionId;
function isUnqualifiedRevocationRegistryId(revocationRegistryId) {
    return exports.unqualifiedRevocationRegistryIdRegex.test(revocationRegistryId);
}
exports.isUnqualifiedRevocationRegistryId = isUnqualifiedRevocationRegistryId;
function isUnqualifiedSchemaId(schemaId) {
    return exports.unqualifiedSchemaIdRegex.test(schemaId);
}
exports.isUnqualifiedSchemaId = isUnqualifiedSchemaId;
function isDidIndySchemaId(schemaId) {
    return exports.didIndySchemaIdRegex.test(schemaId);
}
exports.isDidIndySchemaId = isDidIndySchemaId;
function isDidIndyCredentialDefinitionId(credentialDefinitionId) {
    return exports.didIndyCredentialDefinitionIdRegex.test(credentialDefinitionId);
}
exports.isDidIndyCredentialDefinitionId = isDidIndyCredentialDefinitionId;
function isDidIndyRevocationRegistryId(revocationRegistryId) {
    return exports.didIndyRevocationRegistryIdRegex.test(revocationRegistryId);
}
exports.isDidIndyRevocationRegistryId = isDidIndyRevocationRegistryId;
function parseIndyDid(did) {
    const match = did.match(exports.didIndyRegex);
    if (match) {
        const [, namespace, namespaceIdentifier] = match;
        return { namespace, namespaceIdentifier };
    }
    else {
        throw new core_1.AriesFrameworkError(`${did} is not a valid did:indy did`);
    }
}
exports.parseIndyDid = parseIndyDid;
function parseIndySchemaId(schemaId) {
    const didIndyMatch = schemaId.match(exports.didIndySchemaIdRegex);
    if (didIndyMatch) {
        const [, did, namespace, namespaceIdentifier, schemaName, schemaVersion] = didIndyMatch;
        return {
            did,
            namespaceIdentifier,
            schemaName,
            schemaVersion,
            namespace,
        };
    }
    const legacyMatch = schemaId.match(exports.unqualifiedSchemaIdRegex);
    if (legacyMatch) {
        const [, did, schemaName, schemaVersion] = legacyMatch;
        return {
            did,
            namespaceIdentifier: did,
            schemaName,
            schemaVersion,
        };
    }
    throw new Error(`Invalid schema id: ${schemaId}`);
}
exports.parseIndySchemaId = parseIndySchemaId;
function parseIndyCredentialDefinitionId(credentialDefinitionId) {
    const didIndyMatch = credentialDefinitionId.match(exports.didIndyCredentialDefinitionIdRegex);
    if (didIndyMatch) {
        const [, did, namespace, namespaceIdentifier, schemaSeqNo, tag] = didIndyMatch;
        return {
            did,
            namespaceIdentifier,
            schemaSeqNo,
            tag,
            namespace,
        };
    }
    const legacyMatch = credentialDefinitionId.match(exports.unqualifiedCredentialDefinitionIdRegex);
    if (legacyMatch) {
        const [, did, schemaSeqNo, tag] = legacyMatch;
        return {
            did,
            namespaceIdentifier: did,
            schemaSeqNo,
            tag,
        };
    }
    throw new Error(`Invalid credential definition id: ${credentialDefinitionId}`);
}
exports.parseIndyCredentialDefinitionId = parseIndyCredentialDefinitionId;
function parseIndyRevocationRegistryId(revocationRegistryId) {
    const didIndyMatch = revocationRegistryId.match(exports.didIndyRevocationRegistryIdRegex);
    if (didIndyMatch) {
        const [, did, namespace, namespaceIdentifier, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag] = didIndyMatch;
        return {
            did,
            namespaceIdentifier,
            schemaSeqNo,
            credentialDefinitionTag,
            revocationRegistryTag,
            namespace,
        };
    }
    const legacyMatch = revocationRegistryId.match(exports.unqualifiedRevocationRegistryIdRegex);
    if (legacyMatch) {
        const [, did, schemaSeqNo, credentialDefinitionTag, revocationRegistryTag] = legacyMatch;
        return {
            did,
            namespaceIdentifier: did,
            schemaSeqNo,
            credentialDefinitionTag,
            revocationRegistryTag,
        };
    }
    throw new Error(`Invalid revocation registry id: ${revocationRegistryId}`);
}
exports.parseIndyRevocationRegistryId = parseIndyRevocationRegistryId;
